doc/source/OEP/OEP_002.rst

```diff
@@ -0,0 +1,239 @@
+===================
+ OEP 2 : read-mode
+===================
+
+
+.. contents::
+   :local:
+
+Status
+======
+
+**Discussion**
+
+
+Branches and Pull requests
+==========================
+
+
+Abstract
+========
+
+This OEP proposes to add a 'read mode' to ophyd devices.  Ophyd
+devices may be read in different modes that we want to access with in
+the same 'stage' context.  For example, one could imagine a streaming
+mode where an ophyd object upon calling its
+``obj.read(read_mode='streaming')`` immediately after triggering a
+detector to get an event which can be consumed 'live' and then later
+calling ``obj.read()`` after the detector has finished.  The two reads
+may contain different keys (as not all children may support streaming).
+
+A second example would be a detector which requires dark frames to be
+taken and stored.  This could be implemented via
+``obj.read(read_mode='dark')`` which would return the dark field. and
+``obj.read()`` would return the bright field.
+
+Another example is a detector with a scalar and array mode where
+``obj.read(read_mode='scalar')`` would just grab the current value and
+``obj.read()`` would read everything.
+
+
+Detailed description
+====================
+
+Changes to function signatures
+------------------------------
+
+To support this change we will have to add an optional keyword
+argument ``read_mode`` to ``trigger``, ``read``, and ``describe``.  If
+a device accepts the kwarg (and a particular value of the kwarg) on
+any one of the three methods then it must accept it on all of them.
+The values accepted for ``read_mode`` should be strings.
+
+Objects much have a 'default' read mode so ``read()`` works.
+
+``trigger`` may dis-regard ``read_mode``, but must accept it if
+``read`` does.
+
+An object may require a matched ``trigger(read_mode=FOO)`` to go with
+a ``read(read_mode==FOO)``.
+
+Changes to the RunEngine
+------------------------
+
+The descriptor cache will now need to be by keyed by ``(obj,
+msg.kwargs.get('read_mode', None))`` rather than just ``obj`` and the
```

Comment:
My initial feeling was to hash obj and `stream_name` in the descriptor cache. This is much better, I like it (the hash identifying the object should know *nothing* of the stream). Great catch!  :+1: